File: hello-mcp-csharp/mcp-client/OllamaClient.cs
using Microsoft.Extensions.Logging;
using System.Net.Http.Json;
using System.Text.Json;
using System.Text.Json.Serialization;

namespace Feuyeux.AI.Hello.Client;

/// <summary>
/// 消息类
/// </summary>
public class Message
{
    [JsonPropertyName("role")]
    public string Role { get; set; } = string.Empty;

    [JsonPropertyName("content")]
    public string Content { get; set; } = string.Empty;

    public Message() { }

    public Message(string role, string content)
    {
        Role = role;
        Content = content;
    }
}

/// <summary>
/// 工具调用类
/// </summary>
public class ToolCall
{
    [JsonPropertyName("function")]
    public FunctionCall Function { get; set; } = new();
}

public class FunctionCall
{
    [JsonPropertyName("name")]
    public string Name { get; set; } = string.Empty;

    [JsonPropertyName("arguments")]
    public Dictionary<string, object> Arguments { get; set; } = new();
}

/// <summary>
/// 聊天响应类
/// </summary>
public class ChatResponse
{
    [JsonPropertyName("message")]
    public ResponseMessage Message { get; set; } = new();
}

public class ResponseMessage
{
    [JsonPropertyName("role")]
    public string Role { get; set; } = string.Empty;

    [JsonPropertyName("content")]
    public string Content { get; set; } = string.Empty;

    [JsonPropertyName("tool_calls")]
    public List<ToolCall>? ToolCalls { get; set; }

    public bool HasToolCalls() => ToolCalls != null && ToolCalls.Count > 0;
}

/// <summary>
/// Ollama 客户端
/// 
/// 用于与 Ollama API 交互，支持工具调用
/// </summary>
public class OllamaClient : IDisposable
{
    private readonly ILogger<OllamaClient> _logger;
    private readonly HttpClient _httpClient;
    private readonly string _baseUrl;
    private readonly string _model;

    public OllamaClient(ILogger<OllamaClient> logger, string baseUrl = "http://localhost:11434", string model = "qwen2.5:latest")
    {
        _logger = logger;
        _baseUrl = baseUrl;
        _model = model;
        _httpClient = new HttpClient
        {
            Timeout = TimeSpan.FromSeconds(300)
        };
    }

    /// <summary>
    /// 发送聊天请求
    /// </summary>
    public async Task<ResponseMessage> Chat(List<Message> messages, List<object> tools)
    {
        try
        {
            _logger.LogInformation("发送聊天请求到 Ollama: model={Model}, messages={Count}", _model, messages.Count);

            var requestBody = new
            {
                model = _model,
                stream = false,
                messages,
                tools
            };

            var json = JsonSerializer.Serialize(requestBody, new JsonSerializerOptions
            {
                WriteIndented = true,
                DefaultIgnoreCondition = JsonIgnoreCondition.WhenWritingNull
            });
            _logger.LogDebug("请求 JSON: {Json}", json);

            var response = await _httpClient.PostAsJsonAsync($"{_baseUrl}/api/chat", requestBody);

            _logger.LogDebug("响应状态: {StatusCode}", response.StatusCode);

            if (!response.IsSuccessStatusCode)
            {
                var errorContent = await response.Content.ReadAsStringAsync();
                _logger.LogError("Ollama API 请求失败: {StatusCode}, {Content}", response.StatusCode, errorContent);
                throw new HttpRequestException($"Ollama API 请求失败: {response.StatusCode}");
            }

            var chatResponse = await response.Content.ReadFromJsonAsync<ChatResponse>();
            if (chatResponse == null)
            {
                throw new InvalidOperationException("无法解析 Ollama 响应");
            }

            return chatResponse.Message;
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Ollama 请求失败");
            throw;
        }
    }

    /// <summary>
    /// 执行工具调用
    /// </summary>
    public async Task<string> ExecuteToolCall(ToolCall toolCall, HelloClient helloClient)
    {
        try
        {
            _logger.LogInformation("执行工具调用: {Name}, 参数: {Arguments}", 
                toolCall.Function.Name, 
                JsonSerializer.Serialize(toolCall.Function.Arguments));

            string result;
            if (toolCall.Function.Name == "get_element")
            {
                // Ollama 可能返回 "name" 或 "element_name"
                var name = "";
                if (toolCall.Function.Arguments.TryGetValue("name", out var nameObj))
                {
                    name = nameObj.ToString() ?? "";
                }
                else if (toolCall.Function.Arguments.TryGetValue("element_name", out var elementNameObj))
                {
                    name = elementNameObj.ToString() ?? "";
                }
                
                if (string.IsNullOrEmpty(name))
                {
                    result = JsonSerializer.Serialize(new { error = "缺少参数: name 或 element_name" });
                }
                else
                {
                    result = await helloClient.GetElement(name);
                }
            }
            else if (toolCall.Function.Name == "get_element_by_position")
            {
                if (!toolCall.Function.Arguments.TryGetValue("position", out var positionObj))
                {
                    result = JsonSerializer.Serialize(new { error = "缺少参数: position" });
                }
                else
                {
                    var position = positionObj is JsonElement jsonElement 
                        ? jsonElement.GetInt32() 
                        : Convert.ToInt32(positionObj);
                    result = await helloClient.GetElementByPosition(position);
                }
            }
            else
            {
                result = JsonSerializer.Serialize(new { error = $"未知工具: {toolCall.Function.Name}" });
            }

            _logger.LogInformation("工具调用结果: {Result}", result);
            return result;
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "工具调用失败");
            return JsonSerializer.Serialize(new { error = ex.Message });
        }
    }

    public void Dispose()
    {
        _httpClient?.Dispose();
    }
}

File: hello-mcp-go/mcp-client/ollama_client.go
package main

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"log"
	"net/http"
	"time"
)

// Message 消息类
type Message struct {
	Role    string `json:"role"`
	Content string `json:"content"`
}

// ToolCall 工具调用类
type ToolCall struct {
	Function struct {
		Name      string                 `json:"name"`
		Arguments map[string]interface{} `json:"arguments"`
	} `json:"function"`
}

// ChatResponse 聊天响应类
type ChatResponse struct {
	Message struct {
		Role      string     `json:"role"`
		Content   string     `json:"content"`
		ToolCalls []ToolCall `json:"tool_calls,omitempty"`
	} `json:"message"`
}

// HasToolCalls 检查是否有工具调用
func (r *ChatResponse) HasToolCalls() bool {
	return len(r.Message.ToolCalls) > 0
}

// OllamaClient Ollama 客户端
type OllamaClient struct {
	BaseURL    string
	Model      string
	HTTPClient *http.Client
}

// NewOllamaClient 创建新的 Ollama 客户端
func NewOllamaClient(baseURL, model string) *OllamaClient {
	if baseURL == "" {
		baseURL = "http://localhost:11434"
	}
	if model == "" {
		model = "qwen2.5:latest"
	}
	return &OllamaClient{
		BaseURL: baseURL,
		Model:   model,
		HTTPClient: &http.Client{
			Timeout: 300 * time.Second,
		},
	}
}

// Chat 发送聊天请求
func (c *OllamaClient) Chat(ctx context.Context, messages []Message, tools []map[string]interface{}) (*ChatResponse, error) {
	log.Printf("发送聊天请求到 Ollama: model=%s, messages=%d", c.Model, len(messages))

	// 构建请求体
	requestBody := map[string]interface{}{
		"model":    c.Model,
		"stream":   false,
		"messages": messages,
		"tools":    tools,
	}

	jsonData, err := json.Marshal(requestBody)
	if err != nil {
		return nil, fmt.Errorf("序列化请求失败: %w", err)
	}

	// 发送请求
	req, err := http.NewRequestWithContext(ctx, "POST", c.BaseURL+"/api/chat", bytes.NewBuffer(jsonData))
	if err != nil {
		return nil, fmt.Errorf("创建请求失败: %w", err)
	}
	req.Header.Set("Content-Type", "application/json")

	resp, err := c.HTTPClient.Do(req)
	if err != nil {
		return nil, fmt.Errorf("发送请求失败: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		body, _ := io.ReadAll(resp.Body)
		return nil, fmt.Errorf("Ollama API 请求失败: %d, %s", resp.StatusCode, string(body))
	}

	// 解析响应
	var chatResp ChatResponse
	if err := json.NewDecoder(resp.Body).Decode(&chatResp); err != nil {
		return nil, fmt.Errorf("解析响应失败: %w", err)
	}

	return &chatResp, nil
}

// ExecuteToolCall 执行工具调用
func (c *OllamaClient) ExecuteToolCall(ctx context.Context, toolCall ToolCall, helloClient *HelloClient) (string, error) {
	log.Printf("执行工具调用: %s, 参数: %v", toolCall.Function.Name, toolCall.Function.Arguments)

	var result string
	var err error

	switch toolCall.Function.Name {
	case "get_element":
		name, ok := toolCall.Function.Arguments["name"].(string)
		if !ok {
			return "", fmt.Errorf("参数 name 类型错误")
		}
		result, err = helloClient.GetElement(ctx, name)

	case "get_element_by_position":
		var position int
		switch v := toolCall.Function.Arguments["position"].(type) {
		case float64:
			position = int(v)
		case int:
			position = v
		default:
			return "", fmt.Errorf("参数 position 类型错误")
		}
		result, err = helloClient.GetElementByPosition(ctx, position)

	default:
		return "", fmt.Errorf("未知工具: %s", toolCall.Function.Name)
	}

	if err != nil {
		log.Printf("工具调用失败: %v", err)
		errorJSON, _ := json.Marshal(map[string]string{"error": err.Error()})
		return string(errorJSON), nil
	}

	log.Printf("工具调用结果: %s", result)
	return result, nil
}
